### ISSUE TITLE
PR #8959 breaks IPv6 support

### ISSUE THREAD

**Issue by mathias-ioki** [grafana/loki#9462]
**Describe the bug**
In the PR https://github.com/grafana/loki/pull/8959 the `memberlist.advertise_addr` is set to `common.instance_addr` if empty:

```
r.MemberlistKV.AdvertiseAddr = r.Common.InstanceAddr
```

The problem is, that these blocks need different formats for the IPv6 address. The `common` block needs the IPv6 address with brackets (`[::1]`) and the memberlist block expects the IPv6 address without brackets (`::1`). When we now copy the IPv6 address from the common section to the memberlist section, we copy also the brackets and memberlist starts to complain about it.

**To Reproduce**
* Configure an IPv6 `instance_addr` in the common section:
```
common:
  instance_addr: "[::1]"
```
* Don't configure `advertis_addr` in memberlist section.

**Environment:**
 - Loki version: 2.8.2
 - Loki runs in `simple scalable deployment mode` as service in a VM
 - OS: Ubuntu 22.04.2 LTS

**Log message**
```
failed: failed to create memberlist: Failed to get final advertise address: failed to parse advertise address \"[::1]\""
```

**Comment by periklis** [comment]
I agree this is a small hindsight I missed because of Loki's unique common config block when upgrading dskit to include the grafana/dskit#185 IPv6 support. Would you like to support us with a small contribution aka a fix?

**Comment by leahoswald** [comment]
Run into this problem after an upgrade. Would be great so see a fix soon.

**Comment by frittentheke** [comment]
> I agree this is a small hindsight I missed because of Loki's unique common config block when upgrading dskit to include the grafana/dskit#185 IPv6 support. Would you like to support us with a small contribution aka a fix?

@periklis what would be your idea for a fix?
A change to Loki's config parsing / validation?

**Comment by periklis** [comment]
@frittentheke Yes probably either a fix in the [config_wrapper](https://github.com/grafana/loki/blob/main/pkg/loki/config_wrapper.go#L169) or deeper down to the `r.MemberListKV` that transforms from one format to the other. My gut feeling says later is a better place to start on.

**Comment by flokli** [comment]
I did dive into this.

`AdvertiseAddr` and `AdvertisePort` fields in `memberlist.Config` are simply passed down in the `memberlist.Transport` `FinalAdvertiseAddr` impls.

There's no need to put IPv6 addresses in brackets if they're not used in a "sockaddr context" (ip and port), so golangs `net.ParseIP` in `vendor/github.com/hashicorp/memberlist/net_transport.go` (rightfully) refuses to parse `[::1]` as an IP.

As a conclusion, specifying the IP without the brackets:

```
common:
  instance_addr: "::1"
```

does work.

IMHO, the behaviour is correct here, `instance_addr` doesn't accept ports (that's what `instance_port` is for), so there shouldn't be brackets around raw IPv6 addrs.

There could be some logic, detecting the "sockaddr without port" case and unwrapping it while logging a warning, but I'm not sure it's worth it and could rather be fixed by providing better documentation on the IPv6 cases.
